File: Data Structures & Algorithms/longest-increasing-subsequence/submission-3.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        /*
            9   1   4   2   3   3   7
            1   1   2   2   3   1   1
                            |
        */
        int n = nums.size();
        vector<int>dp(n,1);        
        for(int i=1; i<n; i++){
            for(int j=0; j<i; j++){
                if(nums[i] > nums[j]){
                    dp[i] = max(dp[i],1+dp[j]);
                }
            }
        }
        return *max_element(dp.begin(),dp.end());
    }
};

File: Data Structures & Algorithms/longest-consecutive-sequence/submission-3.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        """
            subsequence

            2   20  4   10  3   4   5


            2:  1
            20: 1
            4:  1
            10: 1
            3:  1
            4:  1
            5:  1

        """

        n = len(nums)
        mp = defaultdict(int)
        result = 0

        for num in nums:
            mp[num] = 1

        for k in sorted(mp):
            if k-1 in mp:
                mp[k] += mp[k-1]
            
            result = max(result,mp[k])

        return result